molecularnodes/io/universe/universe.py

```diff
@@ -0,0 +1,517 @@
+                "value": self.is_lipid,
+                "type": "BOOLEAN",
+                "domain": "POINT",
+            },
+            "is_peptide": {
+                "value": self.is_peptide,
+                "type": "BOOLEAN",
+                "domain": "POINT",
+            },
+        }
+
+    def create_model(
+        self,
+        style: str = "vdw",
+        name: str = "NewUniverseObject",
+        subframes: int = 0,
+        # in_memory: bool = False,
+    ):
+        bob = obj.create_object(
+            name=name, collection=coll.mn(), vertices=self.positions, edges=self.bonds
+        )
+        self.object = bob
+
+        for att_name, att in self._attributes_2_blender.items():
+            try:
+                obj.set_attribute(
+                    bob, att_name, att["value"], att["type"], att["domain"]
+                )
+            except Exception as e:
+                print(e)
+
+        if hasattr(self.atoms, "segindices"):
+            segs = []
+            for seg in self.atoms.segments:
+                segs.append(seg.atoms[0].segid)
+
+            bob["segments"] = segs
+
+        bob["chain_ids"] = self.chain_ids
+        bob["atom_type_unique"] = self.atom_type_unique
+        bob.mn.subframes = subframes
+        bob.mn.molecule_type = "md"
+
+        if style is not None:
+            nodes.create_starting_node_tree(bob, style=style, name=f"MN_{bob.name}")
+
+        bpy.context.view_layer.objects.active = bob
+        bob.mn.uuid = self.uuid
+
+        return bob
+
+    @persistent
+    def _update_trajectory(self, frame):
+        """
+        The function that will be called when the frame changes.
+        It will update the positions and selections of the atoms in the scene.
+        """
+        universe = self.universe
+        frame_mapping = self.frame_mapping
+        bob = self.object
+        if bob is None:
+            raise ObjectMissingError(
+                "Object is deleted and unable to establish a connection with a new Blender Object."
+            )
+        try:
+            subframes = bob.mn.subframes
+            interpolate = bob.mn.interpolate
+        except ReferenceError as e:
+            print(e)
+            return None
+
+        if frame < 0:
+            return None
+
+        if frame_mapping:
+            # add the subframes to the frame mapping
+            frame_map = np.repeat(frame_mapping, subframes + 1)
+            # get the current and next frames
+            frame_a = frame_map[frame]
+            frame_b = frame_map[frame + 1]
+
+        else:
+            # get the initial frame
+            if subframes == 0:
+                frame_a = frame
+            else:
+                frame_a = int(frame / (subframes + 1))
+
+            # get the next frame
+            frame_b = frame_a + 1
+
+        if frame_a >= universe.trajectory.n_frames:
+            return None
+
+        # set the trajectory at frame_a
+        universe.trajectory[frame_a]
```

Comment:
Reloading the trajectory timestep (`universe.trajectory[frame]`) can negatively impact performance because it reads the data from the file and overwrites the current frame, even if it's already at that frame. A solution is to check the current frame before reloading:

```python

@staticmethod
def _goto_frame(universe, frame):
    if universe.trajectory.frame != frame:
        universe.trajectory[frame]
```

Reply:
More generally, I'm considering the possibility of implementing caching for these operations. I have tried the following approach:

```python
from functools import lru_cache

@lru_cache(maxsize=10)
def _goto_frame(self, frame):
    if self.universe.trajectory.frame != frame:
        self.universe.trajectory[frame]
    return self.positions
```

This method caches the 10 most recent outputs, avoiding the need to repeatedly reload the trajectory.

However, this approach results in different visualizations with interpolation. I'll need to investigate further to understand why. If you can figure it out, it would be great!

Reply:
FYI, by performance, I mean the frame rate. If I switch the frame rate to 240 fps, the actual frame rate without subframe processing is 90 fps, and it drops to 60 fps with the old implementation.

Reply:
Thanks for the tip - I didn't realise that it would reprocess the frame if it was already on it. I don't know how beneficial frame caching would be - users are usually playing back continuously I would imagine? Potential to implement, but maybe in a later PR rather than for the 4.2 deadlin (july 15) I am pushing for.

Reply:
> Potential to implement, but maybe in a later PR rather than for the 4.2 deadlin (july 15) I am pushing for.

Yes, I completely agree.

I was mainly thinking about the potential cost of I/O operations when we want to stream trajectories from e.g. AWS in the future.

Reply:
For streaming this would definitely make sense!